# Spam Detector

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [2]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

# Comparing Models for Spam Detection: Logistic Regression vs. Random Forest

## Why Random Forest is Better for Spam Detection

- **Handles Non-Linear Relationships:**  
  Spam detection often involves complex patterns in email text, metadata, and sender behaviors. Random Forests excel at capturing these non-linear relationships.

- **Feature Importance:**  
  Random Forest can identify which features (e.g., certain words, phrases, or email metadata) contribute most to classifying emails as spam or not spam, which is helpful for further refining the model.

- **Robust to Noisy Data:**  
  Emails may contain noisy, irrelevant, or unpredictable patterns (e.g., typos, random characters in spam emails). Random Forest is robust to such noise.

- **Versatility:**  
  Random Forest can work well with diverse types of features (e.g., text embeddings, numerical metadata like "number of recipients," categorical features like "sender domain").

- **Outlier Handling:**  
  Spam emails often behave like outliers compared to normal emails. Random Forest is inherently good at detecting such anomalies.

---

## When Logistic Regression Might Work for Spam Detection

- **Interpretability:**  
  If it’s critical to understand why an email is classified as spam (e.g., for legal compliance or user trust), Logistic Regression provides clear coefficients for each feature.  
  *Example:* "The presence of the word 'free' increases the likelihood of spam by 80%."

- **Computational Simplicity:**  
  If the dataset is small or the classification problem doesn’t require complex modeling, Logistic Regression can be a simple and effective choice.

- **Baseline Performance:**  
  Logistic Regression can be used as a baseline model. It works surprisingly well when you use preprocessed text features like TF-IDF or bag-of-words.







## Split the Data into Training and Testing Sets

In [3]:
# Create the labels set `y` and features DataFrame `X`
y = data['spam']  # Labels: Spam (1) or Not Spam (0)
X = data.drop(columns=['spam'])  # Features: All columns except 'spam'

# Display the shapes of X and y
print("Features (X) shape:", X.shape)
print("Labels (y) shape:", y.shape)

Features (X) shape: (4601, 57)
Labels (y) shape: (4601,)


In [4]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
label_balance = y.value_counts()

# Display the counts of each class
print("Label Balance:")
print(label_balance)

Label Balance:
spam
0    2788
1    1813
Name: count, dtype: int64


In [6]:
# Split the data into X_train, X_test, y_train, y_test
from sklearn.model_selection import train_test_split
import pandas as pd

# Split the data into training and testing sets (80% train, 20% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

# Create a summary table
split_summary = pd.DataFrame({
    "Dataset": ["X_train", "X_test", "y_train", "y_test"],
    "Rows": [X_train.shape[0], X_test.shape[0], y_train.shape[0], y_test.shape[0]],
    "Columns": [X_train.shape[1], X_test.shape[1], 1, 1]
})

# Display the summary table
print(split_summary)


   Dataset  Rows  Columns
0  X_train  3680       57
1   X_test   921       57
2  y_train  3680        1
3   y_test   921        1


## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [15]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [9]:
# Fit the Standard Scaler with the training data
scaler.fit(X_train)

StandardScaler()

In [10]:
# Scale the training data
X_train_scaled = scaler.transform(X_train)


## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. Use a `random_state` of `1`. 

In [13]:
# Create and fit a Logistic Regression model
from sklearn.linear_model import LogisticRegression
# Create a Logistic Regression model instance
logistic_model = LogisticRegression(random_state=1)

# Fit the model with the scaled training data
logistic_model.fit(X_train_scaled, y_train)



LogisticRegression(random_state=1)

In [17]:
# Make and save testing predictions with the trained Logistic Regression model using the test data
y_test_predictions = logistic_model.predict(X_test_scaled)
# Review the predictions
# Save the predictions to a DataFrame
predictions_df = pd.DataFrame({
    "Actual": y_test.values,
    "Predicted": y_test_predictions
})

# Review the predictions
predictions_df.head()

Actual  Predicted
0       1          1
1       1          1
2       0          0
3       1          1
4       0          0

In [24]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy = accuracy_score(y_test, y_test_predictions)

print(f"Logistic Regression Accuracy Score: {accuracy}")

Logistic Regression Accuracy Score: 0.9294245385450597


## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. Use a `random_state` of `1`. 

In [20]:
# Create and fit a Random Forest Classifier model
from sklearn.ensemble import RandomForestClassifier
random_forest_model = RandomForestClassifier(random_state=1)
# Fit the model to the scaled training data
random_forest_model.fit(X_train_scaled, y_train)

RandomForestClassifier(random_state=1)

In [23]:
# Make and save testing predictions with the trained Random Forest Classifier model using the test data
rf_test_predictions = random_forest_model.predict(X_test_scaled)
# Review the predictions
# Save the predictions to a DataFrame
rfpredictions_df = pd.DataFrame({
    "Actual": y_test.values,
    "Predicted": rf_test_predictions
})
rfpredictions_df.head()



Actual  Predicted
0       1          1
1       1          1
2       0          0
3       1          1
4       0          0

In [22]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
rf_accuracy = accuracy_score(y_test, rf_test_predictions)

print(f"Random Forest Classifier Accuracy Score: {rf_accuracy}")

Random Forest Classifier Accuracy Score: 0.9391965255157437


## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

# Which Model is Better: Random Forest Classifier or Logistic Regression?

## Accuracy Scores
- **Random Forest Classifier Accuracy**: 93.92%
- **Logistic Regression Accuracy**: 92.94%

---

## Which Model is Better?
The **Random Forest Classifier** performs better than the **Logistic Regression**, as it achieves a higher accuracy score.

---

## Why Random Forest is Better
1. **Non-Linear Relationships**:
   - Spam detection involves complex, non-linear patterns in the features (e.g., word frequencies, character counts). Random Forest is better at modeling these non-linear relationships, while Logistic Regression assumes a linear relationship between features and the target variable.

2. **Feature Interactions**:
   - Random Forest captures interactions between features naturally, while Logistic Regression does not unless features are manually combined or transformed.

3. **Robustness**:
   - Random Forest is more robust to noise and irrelevant features, which are common in datasets like spam detection, where not all features contribute equally.

4. **High Accuracy**:
   - With only a 1% improvement, Random Forest provides better performance, which can be critical in applications like spam detection where even a small difference can impact the user experience.

---

## When to Use Logistic Regression Instead
- **Interpretability**:
  - Logistic Regression is more interpretable because its coefficients clearly show how each feature contributes to the prediction. If understanding why an email is classified as spam is important, Logistic Regression might be preferred.
  
- **Simplicity and Speed**:
  - Logistic Regression is simpler, faster to train, and computationally efficient, making it suitable for real-time systems or when computational resources are limited.

---

## Conclusion
While **Random Forest Classifier** achieves a better accuracy score and is more suitable for capturing the complexity of spam detection, **Logistic Regression** could still be preferred if interpretability or computational efficiency is a higher priority.

# As I predicted which model performs better, Random Forest does indeed perform better. However, after careful consideration of both models, medium-sized companies where speed and simplicity play a role may opt for Logistic Regression instead